lightning/src/routing/router.rs

```diff
@@ -216,84 +359,200 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
+					if let Some(available_msat) = initial_liquidity_available_msat {
+						initial_liquidity_available_msat = Some(cmp::min(available_msat, htlc_maximum_msat));
+					} else {
+						initial_liquidity_available_msat = Some(htlc_maximum_msat);
+					}
+				}
+
+				match initial_liquidity_available_msat {
+					Some(available_msat) => available_msat,
+					// We assume channels with unknown balance have
+					// a capacity of 0.0025 BTC (or 250_000 sats).
+					None => 250_000 * 1000
+				}
+			});
+
+			// It is tricky to substract $next_hops_fee_msat from available liquidity here.
+			// It may be misleading because we might later choose to reduce the value transferred
+			// over these channels, and the channel which was insufficient might become sufficient.
+			// Worst case: we drop a good channel here because it can't cover the high following
+			// fees caused by one expensive channel, but then this channel could have been used
+			// if the amount being transferred over this path is lower.
+			// We do this for now, but this is a subject for removal.
+			if let Some(available_value_contribution_msat) = available_liquidity_msat.checked_sub($next_hops_fee_msat) {
+
+				// Routing Fragmentation Mitigation heuristic:
+				//
+				// Routing fragmentation across many payment paths increases the overall routing
+				// fees as you have irreducible routing fees per-link used (`fee_base_msat`).
+				// Taking too many smaller paths also increases the chance of payment failure.
+				// Thus to avoid this effect, we require from our collected links to provide
+				// at least a minimal contribution to the recommended value yet-to-be-fulfilled.
+				//
+				// This requirement is currently 5% of the remaining-to-be-collected value.
+				// This means as we successfully advance in our collection,
+				// the absolute liquidity contribution is lowered,
+				// thus increasing the number of potential channels to be selected.
+
+				// Derive the minimal liquidity contribution with a ratio of 20 (5%, rounded up).
+				let minimal_value_contribution_msat: u64 = (recommended_value_msat - already_collected_value_msat + 19) / 20;
+				// Verify the liquidity offered by this channel complies to the minimal contribution.
+				let contributes_sufficient_value = available_value_contribution_msat >= minimal_value_contribution_msat;
+
+				let value_contribution_msat = cmp::min(available_value_contribution_msat, $next_hops_value_contribution);
+				// Includes paying fees for the use of the following channels.
+				let amount_to_transfer_over_msat: u64 = match value_contribution_msat.checked_add($next_hops_fee_msat) {
+					Some(result) => result,
+					// Can't overflow due to how the values were computed right above.
+					None => unreachable!(),
+				};
+
+				//TODO: Explore simply adding fee to hit htlc_minimum_msat
+				if contributes_sufficient_value && amount_to_transfer_over_msat >= $directional_info.htlc_minimum_msat {
```

Comment:
I think `htlc_minimum_msat` isn't covered by test ? Don't get failure for increasing by 1, we should check we return a failure when links on the path advertised enough liquidity but the sent payment is too small? 

```
@@ -445,7 +446,7 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
                                };
 
                                //TODO: Explore simply adding fee to hit htlc_minimum_msat
-                               if contributes_sufficient_value && amount_to_transfer_over_msat >= $directional_info.htlc_minimum_msat {
+                               if contributes_sufficient_value && amount_to_transfer_over_msat >= $directional_info.htlc_minimum_msat + 1 {
                                        let hm_entry = dist.entry(&$src_node_id);
                                        let old_entry = hm_entry.or_insert_with(|| {
                                                // If there was previously no known way to access
```

Reply:
Okay, so I split this check into 2 separate checks. For one of them I wrote a special test, and not it fails as you asked. 

For the second check, it's kind of a headscratcher to craft a test, so perhaps we can leave it for now?

Reply:
Ok hold on, the second case needs a bit more work.

Reply:
Added a test for the second case, while founding this issue: 
```
	// TODO: values over channels should not fall below htlc_minimum_msat.
	fn update_value_and_recompute_fees(&mut self, value_msat: u64) {
```